Agora chat (#2461)
AgoraIO/agora_doc_source: en-US/markdown/agora-chat/Develop/Chat Rooms/agora_chat_chatroom_attributes_rn.md

> Chat room attributes consists of basic attributes (such as room name, room description, and room announcement) and custom attributes. When basic attributes cannot satisfy the business requirements, users can add custom attributes that are synchronized with all chat room members. 

Reviewer: ```suggestion
Chat room attributes consist of basic attributes (such as room name, room description, and room announcement) and custom attributes. When basic attributes cannot satisfy the business requirements, users can add custom attributes that are synchronized with all chat room members. 
```

Reviewer: 这里用的是 room name，下面是 subject，建议保持一致。

Custom attributes can be used to store information such as chat room type, game roles, game status, and host positions. They are stored as key-value maps, and the updates of custom attributes are synchronized with all chat room members.

This page shows how to use the Agora Chat SDK to manage basic and custom attributes of chat rooms in your app.

## Understand the tech

> The Agora Chat SDK provides the `ChatManager` and `ChatRoom` classes for chat room management, which allows you to implement the following features:

Reviewer: ```suggestion
The Agora Chat SDK provides the `ChatManager` and `ChatRoom` classes for chat room management, which allow you to implement the following features:
```

- Retrieve or modify basic attributes of a chat room
- Retrieve custom attributes of a chat room
- Set custom attributes of a chat room
- Remove custom attributes of a chat room

## Prerequisites

Before proceeding, ensure that you meet the following requirements:

- You have initialized the Agora Chat SDK. For details, see [Get Started with rn](./agora_chat_get_started_android?platform=rn).
- You understand the call frequency limit of the Agora Chat APIs supported by different pricing plans as described in [Limitations](./agora_chat_limitation?platform=rn).
- You understand the number of chat rooms supported by different pricing plans as described in [Pricing Plan Details](./agora_chat_plan?platform=rn).

## Implementation

This section introduces how to call the APIs provided by the Agora Chat SDK to implement the features listed above.

### Manage basic chat room attributes

#### Retrieve basic chat room attributes

> All the chat room members can call `fetchChatRoomFromServer` to retrieve the detailed information of the current chat room, including the subject, announcements, description, member type, and admin list. 

Reviewer: ```suggestion
All the chat room members can call `fetchChatRoomFromServer` to retrieve the detailed information of the current chat room, including the name, announcements, description, member type, and admin list. 
```

```typescript
// The fetchPublicChatRoomsFromServer method returns basic attributes including ID, name, description, maximum members, owners, roles, and whether all members are muted.
ChatClient.getInstance()
  .roomManager.fetchPublicChatRoomsFromServer(pageNum, pageSize)
  .then((response) => {
    rollLog(`success: ${response}`);
  })
  .catch((error) => {
    rollLog(`fail: ${error}`);
  });
```

#### Change chat room subject or description

> Only the chat room owner and admin can set and update the chat room subject and description.

Reviewer: ```suggestion
Only the chat room owner and admin can set and update the chat room name and description.
```

```typescript
// The chat room owner and admin call changeChatRoomSubject to modify the chat room subject.
ChatClient.getInstance()
  .roomManager.changeChatRoomSubject(roomId, subject)
  .then((response) => {
    rollLog(`success: ${response}`);
  })
  .catch((error) => {
    rollLog(`fail: ${error}`);
  });

// The chat room owner and admin call changeChatroomDescription to modify the chat room description.
ChatClient.getInstance()
  .roomManager.changeChatRoomDescription(roomId, desc)
  .then((response) => {
    rollLog(`success: ${response}`);
  })
  .catch((error) => {
    rollLog(`fail: ${error}`);
  });
```

#### Retrieve or change chat room announcements

> All the chat room members can retrieve the chat room announcements. 

Reviewer: ```suggestion
All the chat room members can retrieve the chat room announcement. 
```

> Only the chat room owner and admin can set and update the announcements. Once the announcements are updated, all the chat room members receive the `onAnnouncementChanged` callback.

Reviewer: ```suggestion
Only the chat room owner and admin can set and update the announcement. Once the announcement is updated, all the chat room members receive the `onAnnouncementChanged` callback.
```

```typescript

> // Chat room members can call fetchChatRoomAnnouncement to retrieve the chat room announcements.

Reviewer: ```suggestion
// Chat room members can call fetchChatRoomAnnouncement to retrieve the chat room announcement.
```

ChatClient.getInstance()
  .roomManager.fetchChatRoomAnnouncement(roomId)
  .then((response) => {
    rollLog(`success: ${response}`);
  })
  .catch((error) => {
    rollLog(`fail: ${error}`);
  });

> // The chat room owner and admin can call updateChatRoomAnnouncement to set or update the chat room announcements.

Reviewer: ```suggestion
// The chat room owner and admin can call updateChatRoomAnnouncement to set or update the chat room announcement.
```

ChatClient.getInstance()
  .roomManager.updateChatRoomAnnouncement(roomId, announcement)
  .then((response) => {
    rollLog(`success: ${response}`);
  })
  .catch((error) => {
    rollLog(`fail: ${error}`);
  });
```

### Manage custom chat room attributes

#### Retrieve specified or all custom attributes 
All chat room members can call `fetchChatRoomAttributes` to retrieve specified or all custom attributes of the chat room.

```typescript
// Retrieves certain custom attributes by specifying chat room ID and attribute keys. Get all if key is empty.
ChatClient.getInstance()
  .roomManager.fetchChatRoomAttributes(roomId, keys)
  .then((response) => {
    rollLog(`success: ${response}`);
  })
  .catch((error) => {
    rollLog(`fail: ${error}`);
  });
```

#### Set custom attribute

> Chat room members can call `addAttributes` to set one single custom attribute. Use this method to add new custom attributes or update existing attributes that set by yourself. After you successfully call this method, other members in the chat room receive an `onAttributesUpdated` callback. 

Reviewer: ```suggestion
Chat room members can call `addAttributes` to set one single custom attribute. Use this method to add new custom attributes or update existing attributes that are set by yourself. After you successfully call this method, other members in the chat room receive an `onAttributesUpdated` callback. 
```

```typescript
// Sets a custom attribute by specifying chat room ID, attribute key, attribute value, deleteWhenLeft and overwrite. \

> // deleteWhenLeft means to delete the property when leaving the room.

Reviewer: ```suggestion
// deleteWhenLeft means to delete the attribute when leaving the room.
```

> // overwrite means forcibly overwrite properties of the same key.

Reviewer: ```suggestion
// overwrite means forcibly overwrite attributes of the same key.
```
